### PR TITLE
create_embeddings and cluster

### PR DESCRIPTION
**Pull request by imy99**

### CODE REVIEW

File: ahl_health_apps_analysis/pipeline/create_embeddings.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+	'''date in form yyyy-mm-dd'''
+	date = '2022-11-04'
+	details = pd.read_csv(f'outputs/data/preprocessed-description-{date}.csv')
+
+	model = SentenceTransformer("all-mpnet-base-v2")
+
+	# Generate sentence embeddings (might take a few minutes for 1000s of sentences)
+	description_embeddings = np.array(model.encode(details.description.to_list()))
+
+	embedding_ids = details["appId"].iloc[:].to_list()
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/create_embeddings.py]: you don't use this or save it, so may as well delete

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+from sklearn.feature_extraction.text import TfidfTransformer
+from sklearn.feature_extraction.text import TfidfVectorizer
+from sklearn.cluster import KMeans
+
+date = ('2022-11-04')
+details = pd.read_csv(f'outputs/data/preprocessed-description-{date}.csv')
+
+# step 1 - load in word embeddings
+pickle_in = open(f'outputs/data/description-embeddings-{date}.pickle','rb')
+description_embeddings = pickle.load(pickle_in)
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: it might be worth writing a comment to say that the order of `details` is the same order as `description_embeddings`, e.g. index 50 of details talks about the same app as embedding 50.

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+embedding_clustering = reducer_clustering.fit_transform(description_embeddings)
+
+def get_top_tf_idf_words(response, top_n=2, exclude_words=[]):
+    sorted_nzs = np.argsort(response.data)
+    return [word for word in feature_names[response.indices[sorted_nzs]] if word not in exclude_words][-(top_n+1):-1]
+
+tfidf = TfidfVectorizer(stop_words='english')
+
+if __name__ == '__main__':
+# pick cluster type:
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: put imports and functions above the `if __name__ == '__main__':` and everything else after it. i.e. all of
```
date = ('2022-11-04')
details = pd.read_csv(f'outputs/data/preprocessed-description-{date}.csv')

# step 1 - load in word embeddings
pickle_in = open(f'outputs/data/description-embeddings-{date}.pickle','rb')
description_embeddings = pickle.load(pickle_in)

# step 2 - reduce word embeddings to lower dimensions
reducer = umap.UMAP(n_components=2, random_state=1)
embedding = reducer.fit_transform(description_embeddings)

reducer_clustering = umap.UMAP(n_components=50, random_state=1)
embedding_clustering = reducer_clustering.fit_transform(description_embeddings)

tfidf = TfidfVectorizer(stop_words='english')

```
can be copied to after the `if __name__ == '__main__':`

Comment (lizgzil) [reply ahl_health_apps_analysis/pipeline/cluster.py]: and the `create_argparser` function can go above it

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+	def create_argparser():
+
+		parser = ArgumentParser()
+		parser.add_argument('--cluster_type', help="Type of cluster", default="kmeans", type=str)
+		return parser
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: you could also have an argument for number of clusters when you are using kmeans

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+			.mark_circle(size=60)
+			.encode(x="x", y="y", tooltip=["cluster_names", "appId", "summary"], color="cluster_names:N")
+		).interactive()
+
+		fig.save(f'outputs/figures/{date}-hdbscan-cluster.html')
+
+	
+	elif args.cluster_type == "kmeans":
+
+		clusterer = KMeans(n_clusters=15)
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: having the number of clusters when using kmeans as an argument would be useful since you could quickly experiment with it e.g. `clusterer = KMeans(n_clusters=args.n_clusters)`

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+		cluster_names = {i:"-".join(get_top_tf_idf_words(response,2,["app", "apps","blood"])) for i, response in enumerate(X)}
+		df['cluster_names'] = df['cluster'].map(cluster_names)
+		
+		fig = (
+    	alt.Chart(df.reset_index(), width=725, height=725)
+    	.mark_circle(size=60)
+    	.encode(x="x", y="y", tooltip=["cluster_names", "appId", "summary"], color="cluster_names:N")
+		).interactive()
+
+		fig.save(f'outputs/figures/{date}-kmeans-cluster.html')
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: if you wanted to be fancy you could also add the number of clusters to the output name `fig.save(f'outputs/figures/{date}-{args.n_clusters}-kmeans-cluster.html')`

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+			cluster_selection_method="leaf",
+			prediction_data=True,
+		)
+		clusterer.fit(embedding_clustering)
+		logging.info(
+			f"{len(set(clusterer.labels_))} clusters with {sum([bool(x) for x in clusterer.labels_ if int(x) == -1])} unassigned apps"
+		)
+
+		# step 4 - 
+		df = details.copy()
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: I think `df` could have a better name - like `app_details`

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+		radius = 3
+		neigh = KNeighborsClassifier(n_neighbors=radius)
+
+		current_unassigned = len(df[df.cluster == -1].cluster.to_list())
+
+		cluster_list = df.cluster.to_list()
+
+		unassigned_indices = []
+		unassigned_indices.extend([i for i, x in enumerate(cluster_list) if x == -1])
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: i think this is just the same as `unassigned_indices = [i for i, x in enumerate(cluster_list) if x == -1]`?

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+		# step 4 - 
+		df = details.copy()
+		df["x"] = embedding[:, 0]
+		df["y"] = embedding[:, 1]
+		df["cluster"] = [int(x) for x in clusterer.labels_]
+		df["cluster_confidence"] = [str(x) for x in clusterer.probabilities_]
+
+
+		# step 5 -
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: might be worth padding out these "step" comments with a bit more detail `step 5 - assign unassigned apps to their nearest cluster"?

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+		unassigned_indices.extend([i for i, x in enumerate(cluster_list) if x == -1])
+		logging.info(f"{len(unassigned_indices)} apps not in clusters")
+
+		X_list = [x for i, x in enumerate(embedding_clustering) if i not in unassigned_indices]
+		y_list = [i for i in cluster_list if i != -1]
+
+		neigh.fit(X_list, y_list)
+
+		for x in unassigned_indices:
+			df.iat[x, cluster_col] = neigh.predict([embedding_clustering[x]])
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: this is cool!

File: ahl_health_apps_analysis/pipeline/cluster.py [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+		#step 6
+		cluster_description = {}
+		for i, cluster in df.groupby('cluster'):
+		    cluster_description[i] = ' '.join(cluster['description'])
+
+		X = tfidf.fit_transform(cluster_description.values())
+		feature_names = np.array(tfidf.get_feature_names())
+
+		cluster_names = {i:"-".join(get_top_tf_idf_words(response,2,["app", "apps","blood"])) for i, response in enumerate(X)}
```
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: was "blood" not a good word to include?

File: ahl_health_apps_analysis/pipeline/cluster.py (hunk 6 above)
Comment (lizgzil) [inline ahl_health_apps_analysis/pipeline/cluster.py]: it looks like all these lines of code are the same in hbscan and kmeans? if so, have the cluster naming and plot building after the `if else` - that way it only needs to be written out once and means you don't need to edit code in two places if you make changes. I guess you will need another `if/else` right at the end with 
```
if :
  fig.save(f'outputs/figures/{date}-hdbscan-cluster.html')
else:
  fig.save(f'outputs/figures/{date}-kmeans-cluster.html')
 ```

**Review by lizgzil — COMMENTED** [review]
excellent! can't see any bugs, but a few suggestions for ways to improve the code

**Review by lizgzil — APPROVED** [review]
